core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -1082,7 +1082,24 @@ object IO extends Serializable {
    * returning a canceler, which will be used by the runtime to cancel the
    * asynchronous effect if the fiber executing the effect is interrupted.
    */
-  final def async0[E, A](register: (Callback[E, A]) => Async[E, A]): IO[E, A] = new AsyncEffect(register)
+  final def asyncInterrupt[E, A](register: (IO[E, A] => Unit) => Either[Canceler, IO[E, A]]): IO[E, A] = {
```

Comment:
What is the behavior when a `Right` is returned from `register`, particularly when `k` is invoked before `register` returns? Does it start two distinct run loops, one from the continuation, and another from the `Right`?

Reply:
No, there's just one run loop. The old one continues with the continuation.

Reply:
I wonder if `Async.Later` could be changed into a case class that accepts a cancel action.

Reply:
Oh, I see _removing_ the canceler allowed you to clean up a lot of code actually 😁

Reply:
Ah, I think I get it now. If the operation is asynchronous (i.e. `k` is invoked), then you can specify a `Canceler`. If the operation is synchronous (i.e. `k` isn't invoked), just return the `IO`. The doc didn't quite make sense the first couple times around.